iree/compiler/Dialect/Stream/Transforms/LayoutSlices.cpp

```diff
@@ -0,0 +1,328 @@
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "iree/compiler/Utils/IndexSet.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-stream-layout-slices"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Stream {
+namespace {
+
+using Slice = IREE::Stream::ResourcePackOp::Slice;
+
+// Packs slices back-to-back with no aliasing. Useful when debugging to remove
+// the aliasing that makes data breakpoints useless.
+//
+// Slice packed offset SSA values will be updated and start at the given
+// |baseOffset|. Returns |baseOffset| + the total size of the allocation
+// aligned to the requirements of |resourceConfig|.
+static Value packSlicesWithNoAliasing(
+    IREE::Stream::ResourcePackOp packOp, Value baseOffset,
+    ArrayRef<Slice> slices, IREE::Stream::ResourceConfigAttr resourceConfig,
+    IndexSet &indexSet, OpBuilder &builder) {
+  auto loc = packOp.getLoc();
+  int64_t offsetAlignment = resourceConfig.getMinBufferOffsetAlignment();
+  int64_t rangeAlignment = resourceConfig.getMinBufferRangeAlignment();
+
+  Value offset = baseOffset;
+  for (auto &slice : slices) {
+    auto sliceSize = builder.createOrFold<IREE::Util::AlignOp>(
+        loc, slice.dynamicSize, rangeAlignment);
+    slice.packedOffset.replaceAllUsesWith(offset);
+    auto valueToAlign =
+        builder.createOrFold<arith::AddIOp>(loc, offset, sliceSize);
+    offset = builder.createOrFold<IREE::Util::AlignOp>(loc, valueToAlign,
+                                                       offsetAlignment);
+  }
+
+  return builder.createOrFold<IREE::Util::AlignOp>(loc, offset, rangeAlignment);
+}
+
+// Packs a set of statically-sized slices by greedy strip packing.
+//
+// This is the same algorithm used in tflite here:
+// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/simple_memory_arena.cc
+// It's not fantastic and can end up with a significant amount of wastage.
+// They do the packing at runtime and as such care more about performance than
+// we do while doing the packing here offline. For this initial version I
+// wanted to ensure we matched apples/apples their implementation.
+//
+// We should use a set of heuristics and pick the smallest one. There are also
+// some really great papers that have approximations (as all of these are -
+// 2D strip packing is NP-hard) such as
+// https://www.sciencedirect.com/science/article/pii/S0925772113001016 that
+// someone with a brain able to parse mathy papers can try implementing.
```

Comment:
😂 
mmmm computer science / math

I wonder for some of these "insert an algorithm here" transforms if we could distill a sandbox for contributors to play around with:

* 1-3 input programs
* commands to run to get IR, statistics, etc. that are relevant

(I know I'd find that useful, since I go through those motions for each major change I make :P)

Reply:
yeah that'd be awesome - same with partitioning - I've got a few I use but having them standardized would be really helpful - it's easy to make algorithms that are really good at one thing and really bad at everything else :P

Reply:
Filed https://github.com/google/iree/issues/7593 to follow up